app/queries/get_incomplete_reference_applications_ready_to_nudge.rb

```diff
@@ -1,49 +1,35 @@
-      .where.not('candidate.account_locked': true)
-      .where.not('candidate.unsubscribed_from_emails': true)
+      .joins("LEFT OUTER JOIN emails ON emails.application_form_id = application_forms.id AND emails.mailer = 'candidate_mailer' AND emails.mail_template = 'nudge_unsubmitted_with_incomplete_references'")
+      .joins('LEFT OUTER JOIN application_choices ac_primary ON ac_primary.application_form_id = application_forms.id')
+      .joins('LEFT OUTER JOIN course_options ON course_options.id = ac_primary.course_option_id')
+      .joins("LEFT OUTER JOIN courses courses_primary ON courses_primary.id = course_options.course_id AND LOWER(courses_primary.level) = 'primary'")
       .current_cycle
+      # Only applications forms with at least one unsubmitted application choice (inflight)
+      .where('application_choices.status': 'unsubmitted')
+      # Filter out candidates who should not be receiving emails about their accounts
+      .where(candidates: { submission_blocked: false, account_locked: false, unsubscribed_from_emails: false })
+      # have not already seen this message
+      .where(emails: { id: nil })
       .inactive_since(7.days.ago)
-      .with_completion(COMMON_COMPLETION_ATTRS)
-      .has_not_received_email(MAILER, MAIL_TEMPLATE)
-      .and(ApplicationForm
-        .where(science_gcse_completed: true)
-        .or(
-          ApplicationForm.where(
-            'NOT EXISTS (:primary)',
-            primary: ApplicationChoice
-              .select(1)
-              .joins(:course)
-              .where('application_choices.application_form_id = application_forms.id')
-              .where('courses.level': 'primary'),
-          ),
-        ))
-    .and(ApplicationForm
-      .where(efl_completed: true)
-      .or(
-        ApplicationForm.where(
-          "first_nationality IN (#{uk_and_irish})",
-        ),
-      ))
-    .joins(
-      "LEFT OUTER JOIN \"references\" ON \"references\".application_form_id = application_forms.id AND \"references\".feedback_status IN ('feedback_requested', 'feedback_provided')",
-    )
-    .joins(
-      "LEFT OUTER JOIN \"application_choices\" ON \"application_choices\".application_form_id = application_forms.id AND \"application_choices\".status = 'unsubmitted'",
-    )
-    .group('application_forms.id', 'candidate.id')
-    .having('count("references".id) < 2 AND count("application_choices".id) > 0')
+      # They've completed most section, but not references
+      .with_completion(COMMON_COMPLETION_ATTRS_WITHOUT_REFERENCES)
+      .where(references_completed: false)
+      # If they are applying for primary courses, they will also need to have completed a science gcse
+      .where('application_forms.science_gcse_completed = TRUE OR courses_primary.id IS NULL')
+      # Only Candidates who are likely to have a high level of English
+      .where('application_forms.efl_completed = TRUE OR application_forms.first_nationality IN (?)', uk_and_irish_names)
```

Comment:
We need to start to map this important queries and the non indexes being filtered.

I added to my personal notes "Potential slow queries vs Non indexes". Maybe we could centralise these notes somewhere? 

The list below doesn't mean we should add index to all of them but just to start identifying the most common ones.

Anyway, these are the non indexes that we filter:

```
application_forms.recruitment_cycle_year
application_forms.updated_at 
application_choices.status
candidates.submission_blocked
candidates.account_locked
candidates.unsubscribed_from_emails
emails.mailer_template
emails.mailer
application_forms.becoming_a_teacher_completed
application_forms.contact_details_completed
application_forms.degrees_completed
application_forms.english_gcse_completed
application_forms.interview_preferences_completed
application_forms.maths_gcse_completed
application_forms.other_qualifications_completed
application_forms.personal_details_completed
application_forms.safeguarding_issues_completed
application_forms.training_with_a_disability_completed
application_forms.volunteering_completed
application_forms.work_history_completed
application_forms.equality_and_diversity_completed
application_forms.references_completed
application_forms.science_gcse_completed
courses_primary.level
application_forms.efl_completed
application_forms.first_nationality
```

Reply:
We definitely need to do some work on our database!